### ISSUE TITLE
Support for DoH3

### ISSUE THREAD

**Issue by JanSkrabal** [coredns/coredns#5539]
**What would you like to be added**:
Currently, the CoreDNS supports various protocols clients can use to connect to the CoreDNS instance. Among those protocols is a DoH (with HTTP/2).
It might be worth considering adding the support for HTTP/3 as well as it gains popularity.

**Why is this needed**:
We are using CoreDNS as a DNS resolver for Mobile devices all around the world. With the current trends, it seems that 
new versions of operating systems on mobile phones are gradually adding support for DoH3.

Examples
* https://security.googleblog.com/2022/07/dns-over-http3-in-android.html
* It seems that it might be part of the iOS 17

Is this something you'd consider adding to the CoreDNS?

**Comment by chrisohaver** [comment]
> Is [support for DoH3] something you'd consider adding to the CoreDNS?

Sounds like a good feature to me. I'd welcome the addition. @JanSkrabal,  is this something you'd like to contribute, or are you looking for someone else in the community to do it?

**Comment by rdrozhdzh** [comment]
Is DoH3 the same as DoQ (DNS-over-QUIC)?

**Comment by Tantalor93** [comment]
> Is DoH3 the same as DoQ (DNS-over-QUIC)?

I would say yes, it should be the same thing. HTTP/3 was known as HTTP over QUIC, before being renamed. I expect something same to be done also officially for DNS over QUIC as it is currently called by [RFC 9250](https://datatracker.ietf.org/doc/html/rfc9250)

**Comment by Tantalor93** [comment]
Currently standard GO library [does not support HTTP/3](https://github.com/golang/go/issues/32204), but after quick search it seems the go to solution for providing HTTP/3 server would be https://github.com/lucas-clemente/quic-go . I can imagine we built on top of it to provide DNS over HTTP/3

**Comment by mnordhoff** [comment]
> Is DoH3 the same as DoQ (DNS-over-QUIC)?

Not exactly. It's analogous to how DoT and DoH (with HTTP/2) both use TLS, and serve similar purposes, but one has an HTTP layer and implementations don't always support both.

**Comment by FruityWelsh** [comment]
> Is DoH3 the same as DoQ (DNS-over-QUIC)?

no DoH3 has http data ride over QUIC where as DoQ doesn't include the http data. There are some privacy concerns with the extra data provided via http vs directly over QUIC. [AdGuard blogpost more about this](https://adguard.com/en/blog/dns-over-quic.html)

That is to say, supporting both would be awesome, but implementing one doesn't mean the other is implemented.

**Comment by FruityWelsh** [comment]
I think one issue might be the [ietf standard of DoH](https://www.rfc-editor.org/rfc/rfc8484) only specifies http/2 as far as I can find. Not to say it it's not being done already of course.

Relevent discussion on http/3 in go standard lib: https://github.com/golang/go/issues/32204

**Comment by Mo-Fatah** [comment]
Hey @FruityWelsh , Is this issue eligible for the LFX Summer Mentorship ? The project is super interesting to me and I would really love to participate!

**Comment by arkhamHack** [comment]
Hi @FruityWelsh @mnordhoff is this issue available for LFX  mentorship? I'm interested to apply.

**Comment by Filippo125** [comment]
Please close this issue, DoH3 is already supported
